### PR TITLE
Change implementation of user activity records split to match resource presence

### PR DESCRIPTION
**Pull request by Envek**
See https://github.com/gravitational/teleport/pull/34954#pullrequestreview-1768063194

### CODE REVIEW

**Review by zmb3 — APPROVED** [review]

**Comment by zmb3** [comment]
Looks fine to me but lets have @espadolini review next week.

File: lib/usagereporter/teleport/aggregating/service.go [hunk 1]
```diff
@@ -50,28 +50,36 @@ func userActivityReportKey(reportUUID uuid.UUID, startTime time.Time) []byte {
...
 
-	for proto.Size(report) > maxItemSize {
-		if len(report.Records) <= 1 {
-			return nil, trace.LimitExceeded("failed to marshal user activity report within size limit (this is a bug)")
+			recordsTail = append(recordsTail, report.Records[len(report.Records)/2:]...)
+			report.Records = report.Records[:len(report.Records)/2]
 		}
 
-		report.Records = report.Records[:len(report.Records)/2]
+		records = recordsTail
```
Comment (espadolini) [inline lib/usagereporter/teleport/aggregating/service.go:78]: WDYT of doing it like this? It avoids moving data around and it makes it more obvious that we're just repeatedly splitting all the original records slice into reports.

```suggestion
		records = records[len(report.Records):]
```

Comment (Envek) [reply lib/usagereporter/teleport/aggregating/service.go]: Good catch, thanks! Applied your suggestion, also added test to check that no records lost

**Review by espadolini — APPROVED** [review]

File: lib/usagereporter/teleport/aggregating/service_test.go [hunk 2]
```diff
@@ -89,6 +89,29 @@ func TestCRUD(t *testing.T) {
...
+	require.NoError(t, err)
+	require.GreaterOrEqual(t, len(reports), 2)            // some reports were split into two
+	require.GreaterOrEqual(t, len(reports[0].Records), 2) // first report was able to contain a few user activity records
+
+	// reassemble records and ensure that nothing was lost
+	recordsCopy := make([]*prehogv1.UserActivityRecord, 0, recordCount)
+	for _, report := range reports {
+		recordsCopy = append(recordsCopy, report.Records...)
+	}
+	require.ElementsMatchf(t, records, recordsCopy, "some user activity records have been lost during splitting")
```
Comment (zmb3) [inline lib/usagereporter/teleport/aggregating/service_test.go:112]:
```suggestion
	require.Equal(t, records, recordsCopy, "some user activity records have been lost during splitting")
```

`ElementsMatchf` is _really_ slow on large collections.

Comment (Envek) [reply lib/usagereporter/teleport/aggregating/service_test.go]: Good to know. By using this assertion I wanted to say that tested code doesn't promise to keep order of records (however, current implementation does keep order).

**Pull request closed without merging**